### PR TITLE
Effective Preferred Locations

### PR DESCRIPTION
**Pull request by tvaron3**
### Background
Preferred locations is an option set on the cosmos client. It indicates to the SDK in what priority to route requests. 
<ins> Example </ins>
- Account Regions (these are the ones that appear in portal): East US, East US 2, West US, West US 2
- Preferred Locations: East US, West US, West US 2
- Read: Requests will be routed to East US first. If for some reason, the requests fail with a retrieable error or status code, then the requests will be retried on West US. If the requests fail again, then the requests will be retried on West US 2. Note that requests will never go to East US 2 because it is not part of the preferred locations.
- Writes: For single write accounts, writes will go to the write region regardless of preferred locations set. For multi write accounts, writes will go to the preferred locations in order.

### Problem 
Previously to this pr, if no preferred locations were passed into the cosmos client then the SDK would route requests only to the hub region for single write accounts. For multi-write accounts, the SDK would route requests only to the endpoint passed in through the client. This is undesirable because it would be opting out customers by default to an availability feature. 

### Solution
If no preferred locations and the global endpoint are passed into the cosmos client, then the SDK will populate it using the account regions. If a regional endpoint is passed in, the existing behavior will be kept. 

**Note**: This is a breaking change for customers using bounded staleness consistency. 
### Relevant Issue
- https://github.com/Azure/azure-sdk-for-python/issues/39376

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py [hunk 1]
```diff
@@ -387,6 +385,12 @@ def update_location_cache(self, write_locations=None, read_locations=None, enabl
                     self.use_multiple_write_locations,
                 )
 
+        # if preferred locations is empty, we should use the read locations from gateway
+        if len(self.preferred_locations) == 0:
```
Comment (ananth7592) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:389]: nit: If `preferred_locations` is null, len(None) may result in `TypeError`. Won't it?

**Review by ananth7592 — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py (hunk 1 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:389]: Yes good catch will change to it.

**Review by simorenoh — APPROVED** [review]
love the way the test is made as well, thanks Tomas!

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py [hunk 2]
```diff
@@ -481,3 +493,46 @@ def GetLocationalEndpoint(default_endpoint, location_name):
                 return locational_endpoint
 
         return None
+
+    @staticmethod
+    def is_global_endpoint(
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:498]: Can we cache the fact that default endpoint is global / regional as the default endpoint doesn't change?

**Comment by jeet1995** [comment]
> All requests will be routed to East US first. If for some reason, the requests fail with a retriable error or status code, then the requests will be retried on West US. If the requests fail again, then the requests will be retried on West US 2. Note that requests will never go to East US 2 because it is not part of the preferred locations.

I think you need to disambiguate between how reads and writes behave.

**Comment by jeet1995** [comment]
Another aspect is testing - it would be good to add coverage in our retry policy tests when the client isn't configured with preferred locations.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py (hunk 2 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:498]: Will no longer follow this approach for figuring out a global endpoint.

**Comment by tvaron3** [comment]
Will follow this up after atm changes.

File: sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        ([REGION_1, REGION_3], locational_endpoint),
+        ([REGION_1, REGION_2, REGION_3], locational_endpoint)
+    ]
+
+def create_account_endpoints_by_location(global_endpoint: str):
+    locational_endpoints = []
+    for region in ACCOUNT_REGIONS:
+        locational_endpoints.append(_location_cache.LocationCache.GetLocationalEndpoint(global_endpoint, region))
+        _location_cache.LocationCache.GetLocationalEndpoint(global_endpoint, REGION_1)
+    account_endpoints_by_location = OrderedDict()
```
Comment (Parvezkhan0) [inline sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py]: There's an unexpected call to GetLocationalEndpoint inside the loop with REGION_1, unrelated to the loop variable.

File: sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            _global_endpoint_manager._GlobalEndpointManager._GetDatabaseAccountStub = self.original_getDatabaseAccountStub
+            _cosmos_client_connection.CosmosClientConnection._GetDatabaseAccountCheck = self.original_getDatabaseAccountCheck
+        expected_dual_endpoints = []
+
+        # if preferred location set should use that
+        if preferred_location:
+            expected_locations = preferred_location
+        # if client created with regional endpoint preferred locations, only use hub region
+        elif default_endpoint != self.host:
+            expected_locations = ACCOUNT_REGIONS[:1]
```
Comment (Parvezkhan0) [inline sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py]: This assumes "hub region" is always ACCOUNT_REGIONS[0] – i.e., “East US”. This may not always be correct if the order in ACCOUNT_REGIONS changes.

File: sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py (hunk 3 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py]: Thanks for the review ! I removed this method as I changed the behavior and added different tests

File: sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py (hunk 4 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/test/test_effective_preferred_locations.py]: True then the test will fail. The account regions come from our testing account setup.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py [hunk 5]
```diff
@@ -71,7 +72,7 @@ def get_endpoints_by_location(new_locations,
                               writes,
                               use_multiple_write_locations):
     # construct from previous object
-    endpoints_by_location = collections.OrderedDict()
+    regional_routing_context_by_location = collections.OrderedDict()
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py]: Can we add type hint for these local variables ? It would be easier to see why we chose this new name for this variable.

I am sure there would be more local or private variables in the classes with the similar cases, but we can update others in an other PR.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py [hunk 6]
```diff
@@ -431,6 +431,18 @@ def update_location_cache(self, write_locations=None, read_locations=None, enabl
                     self.connection_policy.UseMultipleWriteLocations
                 )
 
+        # if preferred locations is empty and the default endpoint is a global endpoint,
+        # we should use the read locations from gateway as effective preferred locations
+        if self.connection_policy.PreferredLocations:
+            self.effective_preferred_locations = self.connection_policy.PreferredLocations
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:438]: Since `self.connection_policy.PreferredLocations` does not change over time, can we move this to where we call it one time? We don't have to check this over and over.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py [hunk 7]
```diff
@@ -431,6 +431,18 @@ def update_location_cache(self, write_locations=None, read_locations=None, enabl
...
+        elif any(
+                context.get_primary() == self.default_regional_routing_context.get_primary()
+                for context in self.account_read_regional_routing_contexts_by_location.values()
+        ):
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py]: This condition is bit confusing, can we add a helper function, such as `is_default_endpoint_global()` or `is_default_endpoint_regional()` ?

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+                expected_dual_endpoints.append(RegionalRoutingContext(locational_endpoint, locational_endpoint))
+            else:
+                expected_dual_endpoints.append(RegionalRoutingContext(locational_endpoint, default_endpoint))
+
+        read_dual_endpoints = client.client_connection._global_endpoint_manager.location_cache.read_regional_routing_contexts
+        assert read_dual_endpoints == expected_dual_endpoints
+
+    @pytest.mark.cosmosMultiRegion
+    @pytest.mark.parametrize("error", error())
+    def test_no_preferred_locations_with_errors_async(self, setup, error):
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py]: Can we also test failover with write operations?

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        error_lambda = lambda r:FaultInjectionTransport.error_after_delay(
+            0,
+            error
+        )
+        predicate = lambda r: (FaultInjectionTransport.predicate_is_document_operation(r) and
+                               FaultInjectionTransport.predicate_targets_region(r, uri_down))
+        custom_transport.add_fault(predicate,
+                                   error_lambda)
+        fault_setup = self.setup_method_with_custom_transport(custom_transport=custom_transport)
+        fault_cont = fault_setup["col"]
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py]: Please use the full name for variables
```suggestion
        fault_container = fault_setup["col"]
```

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        resp = fault_cont.read_item(item=item_to_read["id"], partition_key=item_to_read[self.partition_key])
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py]: same for this
```suggestion
        response = fault_container.read_item(item=item_to_read["id"], partition_key=item_to_read[self.partition_key])
```

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations_async.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+                expected_dual_endpoints.append(RegionalRoutingContext(locational_endpoint, locational_endpoint))
+            else:
+                expected_dual_endpoints.append(RegionalRoutingContext(locational_endpoint, default_endpoint))
+
+        read_dual_endpoints = client.client_connection._global_endpoint_manager.location_cache.read_regional_routing_contexts
+        assert read_dual_endpoints == expected_dual_endpoints
+
+    @pytest.mark.cosmosMultiRegion
+    @pytest.mark.parametrize("error", error())
+    async def test_no_preferred_locations_with_errors_async(self, setup, error):
```
Comment (tvaron3) [inline sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations_async.py]: Add test using excluded locations and no preferred locations empty

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py (hunk 5 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py]: Added type hint here and all constructors

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py (hunk 6 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:438]: Changed to set this in the constructor

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py (hunk 7 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py]: Moved to a helper function

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py (hunk 8 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py]: added test with write operations

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py (hunk 9 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py]: done

File: sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations_async.py (hunk 11 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations_async.py]: added

File: sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py (hunk 6 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:438]: Actually I realized this could change over time and moving to where it is called once would be a breaking change.

Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_location_cache.py:438]: Makes sense. Thank you for clarifying it.

**Review by allenkim0129 — APPROVED** [review]
LGTM

**Review by kushagraThapar — APPROVED** [review]

**Review by allenkim0129 — APPROVED** [review]
